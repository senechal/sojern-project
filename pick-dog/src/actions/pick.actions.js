import { get } from 'axios';
import {
    NUMBER_OF_DOGS_PER_REQUEST,
    API_URL,
} from '../utils/constants';

import {
    REQUEST_DOG,
    SAVE_DOG,
} from '../utils/actionTypes';

const getDogsImgs = () => {
    const range = [...Array(NUMBER_OF_DOGS_PER_REQUEST)];
    const promises = range.map(() => {
        return get(API_URL);
    });
    return Promise.all(promises);
};
const getActionLifecicle = (action) => ({
    pending: `${action}--pending`,
    success: `${action}--success`,
    failure: `${action}--failure`,
});


const trigger = (type, payload) => ({type, payload});

export const requestDogs = () => dispatch => {
    const {
        pending,
        success,
        failure,
    } = getActionLifecicle(REQUEST_DOG);
    dispatch(trigger(pending));


    return getDogsImgs().then((res) => {
        dispatch(trigger(success, res));
    }).catch(err => {
        dispatch(trigger(failure, err));
    });
}

export const saveImg = (id) => (dispatch, getState) => {
    const { pick: { imgs}} = getState();
    const img = imgs.find(({id: imgId}) => imgId === id);
    if(img){
        dispatch(trigger(SAVE_DOG, {...img}));
    }
}