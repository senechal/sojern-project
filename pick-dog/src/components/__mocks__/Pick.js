const Pick = () => {
    return <div>Pick Mock</div>
};

export default Pick;