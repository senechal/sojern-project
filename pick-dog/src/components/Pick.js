import { useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NotificationManager } from 'react-notifications';
import { requestDogs, saveImg } from "../actions/pick.actions";
import { ButtonContainer } from './styles';
import Galery from './Galery';

const Pick = () => {
    const dispatch = useDispatch();
    const { imgs, loading } = useSelector(state => state.pick);

    const handleClickNext = useCallback(() => {
        dispatch(requestDogs());
    }, [dispatch]);

    const handleSaveImg = useCallback((id) => {
        NotificationManager.success('Image added to Favorites!');
        dispatch(saveImg(id));
    }, [dispatch]);

    useEffect(() => {
        if (!imgs.length){
            dispatch(requestDogs());
        }
    }, [dispatch, imgs]);
    return (
        <>
            <Galery imgs={imgs} loading={loading} onSelect={handleSaveImg} selectable />
            <ButtonContainer>
                <button onClick={handleClickNext} data-testid="pick-button">Next</button>
            </ButtonContainer>
        </>
    )
};

export default Pick;