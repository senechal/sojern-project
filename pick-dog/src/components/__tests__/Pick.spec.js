import { useDispatch, useSelector } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import { requestDogs, saveImg } from '../../actions/pick.actions';
import {
  NUMBER_OF_DOGS_PER_REQUEST,
} from '../../utils/constants';
import Pick from '../Pick';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('../../actions/pick.actions', () => ({
    requestDogs: jest.fn(),
    saveImg: jest.fn(),
}));

describe('Pick tests', () => {

  const dispatch = jest.fn();

  const getImages = () => {
    const range = [...Array(NUMBER_OF_DOGS_PER_REQUEST)];
    return  range.map((v, index) => {
      return { url: `http://mock_image/${index}.jpg`, id: index.toString(), type: 'jpg' }
    });
  }
  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    requestDogs.mockReturnValue('request-dogs-action');
    saveImg.mockReturnValue('save-img-action');
  });

  afterEach(() => {
    dispatch.mockClear();
  });
  it('should, render correctly while loading', () => {
    useSelector.mockReturnValueOnce({loading: true, imgs: []});
    const { container } = render(<Pick />);
    expect(dispatch).toBeCalledWith('request-dogs-action');
    expect(requestDogs).toBeCalled();
    expect(container).toMatchSnapshot();
  });
  it('should, render correctly with imgs', () => {
    useSelector.mockReturnValueOnce({loading: false, imgs: getImages()});
    const { container } = render(<Pick />);
    expect(container).toMatchSnapshot();
  });
  it('should update fotos on button click', () => {
    useSelector.mockReturnValueOnce({loading: false, imgs: getImages()});
    const { getByTestId } = render(<Pick />);
    const button = getByTestId('pick-button');
    fireEvent.click(button);
    expect(dispatch).toBeCalledWith('request-dogs-action');
    expect(requestDogs).toBeCalled();
  });
  it('should save Image to favorites on image click', () => {
    useSelector.mockReturnValueOnce({loading: false, imgs: getImages()});
    const { getAllByTestId } = render(<Pick />);
    const imgs = getAllByTestId('galery-image');
    fireEvent.click(imgs['0']);
    expect(dispatch).toBeCalledWith('save-img-action');
    expect(saveImg).toBeCalledWith('0');
  });
});
