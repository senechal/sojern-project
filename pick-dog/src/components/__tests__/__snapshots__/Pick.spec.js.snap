// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pick tests should, render correctly while loading 1`] = `
<div>
  <div
    class="sc-bdfBwQ bWYwLj"
  >
    <div
      class="sc-hKgILt bcjFbu"
      data-testid="loading-icon"
    />
  </div>
  <div
    class="sc-dlfnbm sc-eCssSg VWpaN"
  >
    <button
      data-testid="pick-button"
    >
      Next
    </button>
  </div>
</div>
`;

exports[`Pick tests should, render correctly with imgs 1`] = `
<div>
  <div
    class="sc-bdfBwQ bWYwLj"
  >
    <img
      alt="http://mock_image/0.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="0"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/0.jpg"
    />
    <img
      alt="http://mock_image/1.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="1"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/1.jpg"
    />
    <img
      alt="http://mock_image/2.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="2"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/2.jpg"
    />
    <img
      alt="http://mock_image/3.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="3"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/3.jpg"
    />
    <img
      alt="http://mock_image/4.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="4"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/4.jpg"
    />
    <img
      alt="http://mock_image/5.jpg"
      class="sc-gsTCUz jhoEbx"
      data-id="5"
      data-testid="galery-image"
      loading="lazy"
      src="http://mock_image/5.jpg"
    />
  </div>
  <div
    class="sc-dlfnbm sc-eCssSg VWpaN"
  >
    <button
      data-testid="pick-button"
    >
      Next
    </button>
  </div>
</div>
`;
